### PR TITLE
Donation endpoints

### LINKED ISSUE #59: Add endpoints for donation
- create/add a new donation #56
- edit a current donation #57
- delete a donation #58

### PR DESCRIPTION
**Pull request by ahmad675**
# Summary
Implemented CRUD functions.
-  GET: (/donations/<user_uuid>)
-  POST: (/create-donation)
-  PUT: (/update_dontation/<donation_uuid>)
-  DELETE: (/<donation_uuid>)

## Test Plan
No test plan yet.

## Closes:
-  #56 (create/add a new donation )
-  #57 (edit a current donation)
-  #58 (delete a donation)
-  #59

### CODE REVIEW

**Review by celinehuang — COMMENTED** [review]
Looks good but it'll be good to write tests for these. Something like this https://github.com/kraglalbert/budgie-web/blob/master/backend/tests/test_transaction_endpoints.py. If you can, add screenshots or video recordings of you sending a request through Postman is good too

**Comment by celinehuang** [comment]
Also I don't believe this code runs?

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:29]: `numTickets` to `num_tickets`

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:30]: `addedBy` to `added_by`

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:41]: here as well

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:39]: `donationSource` to `donation_source`

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:61]: same here

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:67]: here as well

**Review by celinehuang — CHANGES_REQUESTED** [review]
I am changing the field names to follow the Python convention, so you'll need to change these so they will work! (p.s. I might have missed some)

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:72]: @ahmad675 I was wondering whether you could write a unit test to test this? If I recall correctly, if the user were to provide an empty dictionary (e.g. {}) all of the attributes would be `None`, so `name == ""` might not actually check whether it's an empty argument. Or we could check for both tbh.

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:84]: @ahmad675 is the int converter needed here? I think json supports the number type, so the dictionary returned would have `num_tickets` as a number type.

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:93]: @ahmad675 Maybe adding a converter here would help too.

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:96]: @ahmad675 I think you left out `is None` at the end of the if condition?

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:32]: @ahmad675 I was wondering whether you could write a unit test to test this?

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@ahmad675 I was wondering whether you could add unit tests to test your changes in this pr?

@celinehuang After looking at this pr, I wanted to ask you about missing / empty arguments. Aka `{"name" = ""}` vs `{}`. Would we want to check for both?

File: backend/app/donation/routes.py
Comment (celinehuang) [reply backend/app/donation/routes.py:1]: @CurtisMIT hmm good point.. I've never considered that, do you think it would be redundant?

File: backend/app/donation/routes.py
Comment (celinehuang) [reply backend/app/donation/routes.py:9]: hmm this might be referring to get all donations by user so we can sum the total amount.. However, we don't have an amount field yet (task is assigned to @SymPiracha)

Comment (celinehuang) [reply backend/app/donation/routes.py:9]: @CurtisMIT I think we will actually need this for #105, what do you think?

image

Comment (CurtisMIT) [reply backend/app/donation/routes.py:9]: @celinehuang Is MuUser the model for donors?

File: backend/app/donation/routes.py
Comment (aqghawa) [inline backend/app/donation/routes.py:9]: @celinehuang, should I remove this since #105 was closed, and this is no longer necessary?

File: backend/app/donation/routes.py
Comment (celinehuang) [reply backend/app/donation/routes.py:9]: No it is not @CurtisMIT MuUser is just a Mu employee

File: backend/app/donation/routes.py
Comment (celinehuang) [inline backend/app/donation/routes.py:10]: typo with `donations`

File: backend/app/donation/routes.py
Comment (celinehuang) [reply backend/app/donation/routes.py:9]: No @aqghawa we don't need this anymore, thanks fo catching it!

File: backend/app/donation/routes.py
Comment (celinehuang) [reply backend/app/donation/routes.py:25]: https://github.com/hack4impact-mcgill/mu-crm-tool/pull/65/files#r566335911 same issue here where we should be checking for`{}` empty object

**Comment by celinehuang** [comment]
Looks good to me, but I need another pair of eyes so I requested review from @CurtisMIT

File: backend/app/donation/routes.py
Comment (CurtisMIT) [inline backend/app/donation/routes.py:14]: @ahmad675 it's fine to not create an id, as it is automatically created once it is inserted.

**Review by CurtisMIT — CHANGES_REQUESTED** [review]
@ahmad675 tiny thing and it should be good!

**Review by CurtisMIT — APPROVED** [review]
LGTM!

**Review by celinehuang — APPROVED** [review]
🥇

**Pull request merged**
